src/tools/illink/test/Mono.Linker.Tests.Cases/RequiresCapability/RequiresInLibraryAssembly.cs

```diff
@@ -0,0 +1,58 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics.CodeAnalysis;
+using Mono.Linker.Tests.Cases.Expectations.Assertions;
+using Mono.Linker.Tests.Cases.Expectations.Metadata;
+
+namespace Mono.Linker.Tests.Cases.RequiresCapability
+{
+	[SetupLinkerArgument ("-a", "test.exe", "library")]
+
+	[SkipKeptItemsValidation]
+	[ExpectedNoWarnings]
+	public class RequiresInLibraryAssembly
+	{
+		public static void Main ()
+		{
+		}
+
+		[RequiresDynamicCode ("--MethodWhichRequires--")]
+		public static void MethodWhichRequires () { }
+
+		[RequiresDynamicCode ("--InstanceMethodWhichRequires--")]
+		public void InstanceMethodWhichRequires () { }
+	}
+
+	[ExpectedNoWarnings]
+	public sealed class ClassWithDAMAnnotatedMembers
+	{
+		public static void Method ([DynamicallyAccessedMembers (DynamicallyAccessedMemberTypes.PublicMethods)] Type type) { }
+
+		[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods)]
+		public static Type Field;
+	}
+
+	[ExpectedNoWarnings]
+	[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods)]
+	public sealed class ClassWithDAMAnnotation
+	{
+		public void Method () { }
+	}
+
+	[ExpectedNoWarnings]
+	[RequiresUnreferencedCode ("--ClassWithRequires--")]
+	public sealed class ClassWithRequires
```

Comment:
Does the visibility of the class/member matter at all? All of the places I hit this were not public.

Reply:
It can also be internal, but then the assembly must have InternalsVisibleTo - so I intentionally avoided that. It should not matter at all to the warning behavior.

Reply:
You don't need InternalsVisibleTo in order to get the trimmer to emit the bad warning. I hit it here:

https://github.com/dotnet/runtime/pull/84468/files#diff-0feadd0bb2789c2621b7be1a6efe5807907d218fe60d62fd89cc7c4578dbe08cR10

![image](https://user-images.githubusercontent.com/8291187/231251434-fdd6c29c-204e-467b-bb86-d47bab122a31.png)

Reply:
Sort of - it needs IVT if the type is internal... but I guess fields are taken as a whole (probably because of base classes, in which case we should still be able to exclude statics... eventually)

Reply:
So in https://github.com/dotnet/aspnetcore/pull/46518/, this bug was was only hit because the assembly has IVT to tests?